File: ponte/src/spegnimento.js
/* Lo spegnimento programmato del clima (#364): il conto alla rovescia sta
 * nel ponte.
 *
 * «Vorrei uno slider per decidere il tempo che un condizionatore debba
 * restare acceso dal momento che gli do l'on, utile spesso di notte o per
 * accensioni a spot.»
 *
 * La parte che conta non e' lo slider, e' **dove vive il conto alla
 * rovescia**. Un timer nel browser muore chiudendo la pagina, e chi accende
 * il condizionatore per due ore prima di dormire la pagina la chiude sempre:
 * resterebbe acceso tutta la notte. Nella dashboard il timer lo tiene
 * l'integrazione dentro Home Assistant (`spegnimento.py`); qui l'integrazione
 * non c'e' — la plancia la porta l'add-on — e allora lo tiene il ponte, con
 * le stesse tre scelte:
 *
 *   - **la scadenza si scrive sul disco.** Un riavvio dell'add-on nel mezzo
 *     della notte non deve lasciare acceso un condizionatore: all'avvio le
 *     scadenze si rileggono e si riarmano, e quelle gia' passate si eseguono
 *     subito — meglio spegnere in ritardo che non spegnere;
 *   - **lo spegnimento e' `homeassistant.turn_off`.** Nella casella
 *     dell'unita' puo' esserci un `climate`, ma anche uno `switch` o un
 *     `input_boolean`: quel servizio sceglie da se' il dominio giusto;
 *   - **il timer si annulla da solo quando l'unita' viene spenta a mano.**
 *     Chi spegne prima non deve ritrovarsi un timer che, due ore dopo,
 *     spegne un'unita' che nel frattempo qualcun altro aveva riacceso.
 *
 * La plancia chiede tre cose sul filo, con gli stessi nomi dell'integrazione:
 * `dashboardmodern/clima/timer/list`, `/set {entity_id, minuti}` e
 * `/clear {entity_id}`. Chi risponde e' `commissioni.js`; qui c'e' solo il
 * conto.
 */

import { mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";

/* Il piu' corto e il piu' lungo che si possano chiedere, in minuti: sono gli
 * estremi dello slider in plancia. */
export const MIN_MINUTI = 1;
export const MAX_MINUTI = 720;

/* Quanti timer possono stare appesi insieme. Uno per unita' climatica, e le
 * unita' sono una decina: cento e' un fondo, non un limite che si tocca. */
export const MAX_TIMER = 100;

/* Quanto si aspetta prima di scrivere: una raffica di tocchi non e' una
 * raffica di file. */
const SCRIVI_DOPO = 5_000;

const NOME_DEL_FILE = "spegnimenti.json";

export class Spegnimento {
  constructor({
    casa,
    cartella,
    registro = null,
    adesso = () => Date.now(),
    /* I timer si passano da fuori perche' le prove non aspettino: sul banco
     * un'ora dura un istante. */
    programma = (fra, cosa) => setTimeout(cosa, fra),
    annulla = (segno) => clearTimeout(segno),
  } = {}) {
    this.casa = casa;
    this.percorso = cartella ? join(cartella, NOME_DEL_FILE) : "";
    this.registro = registro ?? { info() {}, attenzione() {}, errore() {} };
    this.adesso = adesso;
    this._programma = programma;
    this._annulla = annulla;
    /* entity_id -> scadenza in millisecondi. */
    this._scadenze = new Map();
    /* entity_id -> il segno con cui si disarma l'allarme. */
    this._allarmi = new Map();
    this._scrittura = null;
    this._caricato = false;
    this._lascia = null;
  }

  /* ── memoria ─────────────────────────────────────────────────────────── */

  /* Rilegge le scadenze dal disco e rimette in piedi gli allarmi. Si fa una
   * volta sola; si puo' chiamare quante volte si vuole. */
  carica() {
    if (this._caricato) return;
    this._caricato = true;
    if (this.percorso) {
      try {
        const letto = JSON.parse(readFileSync(this.percorso, "utf8"));
        const grezze = letto?.scadenze;
        if (grezze && typeof grezze === "object") {
          for (const [entita, scadenza] of Object.entries(grezze)) {
            const quando = Number(scadenza);
            if (typeof entita === "string" && entita.includes(".") && Number.isFinite(quando))
              this._scadenze.set(entita, quando);
          }
        }
      } catch (_errore) {
        /* Nessun file, o un file storto: e' come non avere timer. */
      }
    }
    for (const entita of [...this._scadenze.keys()]) this._arma(entita);
    this._ascolta();
  }

  _salva() {
    if (!this.percorso) return;
    if (this._scrittura) return;
    this._scrittura = this._programma(SCRIVI_DOPO, () => {
      this._scrittura = null;
      this.scriviAdesso();
    });
  }

  /* Scrive subito: allo spegnimento dell'add-on, e nelle prove. */
  scriviAdesso() {
    if (!this.percorso) return;
    if (this._scrittura) {
      this._annulla(this._scrittura);
      this._scrittura = null;
    }
    try {
      mkdirSync(dirname(this.percorso), { recursive: true });
      const provvisorio = `${this.percorso}.nuovo`;
      writeFileSync(
        provvisorio,
        JSON.stringify({ scadenze: Object.fromEntries(this._scadenze) }, null, 2),
      );
      renameSync(provvisorio, this.percorso);
    } catch (errore) {
      this.registro.attenzione(`spegnimenti non scritti: ${errore?.message || errore}`);
    }
  }

  /* ── allarmi ─────────────────────────────────────────────────────────── */

  _disarma(entita) {
    const segno = this._allarmi.get(entita);
    if (segno === undefined) return;
    this._allarmi.delete(entita);
    this._annulla(segno);
  }

  /* Mette l'allarme per la scadenza di quell'entita'. Una scadenza gia'
   * passata — il riavvio nel mezzo della notte — non si butta: si esegue
   * subito. */
  _arma(entita) {
    this._disarma(entita);
    const scadenza = this._scadenze.get(entita);
    if (scadenza === undefined) return;
    const fra = Math.max(0, scadenza - this.adesso());
    this._allarmi.set(
      entita,
      this._programma(fra, () => {
        this._allarmi.delete(entita);
        this.spegni(entita).catch((errore) => {
          this.registro.attenzione(
            `${entita}: lo spegnimento programmato non e' riuscito: ${errore?.message || errore}`,
          );
        });
      }),
    );
  }

  /* ── quello che la plancia chiede ────────────────────────────────────── */

  /* Le scadenze appese, cosi' come stanno. */
  scadenze() {
    this.carica();
    return Object.fromEntries(this._scadenze);
  }

  /* Programma lo spegnimento di quell'entita' fra quei minuti. Zero minuti
   * vuol dire «togli il timer», e torna `null` perche' non c'e' nessuna
   * scadenza da mostrare. */
  programma(entita, minuti) {
    this.carica();
    const quanti = Math.trunc(Number(minuti) || 0);
    if (quanti <= 0) {
      this.annulla(entita);
      return null;
    }
    const dentro = Math.max(MIN_MINUTI, Math.min(quanti, MAX_MINUTI));
    if (!this._scadenze.has(entita) && this._scadenze.size >= MAX_TIMER) {
      /* Il fondo: prima cede la scadenza piu' lontana, che e' quella che ha
       * ancora piu' tempo per essere rimessa. */
      let piuLontana = null;
      for (const [quale, quando] of this._scadenze)
        if (piuLontana === null || quando > this._scadenze.get(piuLontana)) piuLontana = quale;
      this._disarma(piuLontana);
      this._scadenze.delete(piuLontana);
    }
    const scadenza = this.adesso() + dentro * 60_000;
    this._scadenze.set(entita, scadenza);
    this._arma(entita);
    this._salva();
    this._ascolta();
    return scadenza;
  }

  /* Toglie il timer di quell'entita', senza toccare l'entita'. */
  annulla(entita) {
    this.carica();
    this._disarma(entita);
    if (this._scadenze.delete(entita)) {
      this._salva();
      this._ascolta();
    }
    return null;
  }

  /* La scadenza e' arrivata: spegni, e togli il timer. */
  async spegni(entita) {
    this._disarma(entita);
    if (this._scadenze.delete(entita)) {
      this._salva();
      this._ascolta();
    }
    await this.casa.chiedi({
      type: "call_service",
      domain: "homeassistant",
      service: "turn_off",
      service_data: { entity_id: entita },
    });
    this.registro.info(`${entita}: spento allo scadere del timer`);
  }

  /* ── chi spegne a mano si porta via anche il timer ───────────────────── */

  /* Segue i cambi di stato finche' c'e' almeno un timer, e nessun altro
   * momento: un'unita' che va a `off` da sola si porta via il suo timer.
   * L'ascolto e' del filo del ponte, che si riapre da solo se cade; se al
   * momento di abbonarsi la casa non risponde, si riprova al prossimo timer. */
  _ascolta() {
    if (!this._scadenze.size) {
      if (this._lascia) {
        const lascia = this._lascia;
        this._lascia = null;
        Promise.resolve()
          .then(() => lascia())
          .catch(() => {});
      }
      return;
    }
    if (this._lascia || this._abbonandomi || typeof this.casa?.ascolta !== "function") return;
    this._abbonandomi = this.casa
      .ascolta("state_changed", (evento) => this._cambiata(evento))
      .then(
        (lascia) => {
          this._abbonandomi = null;
          this._lascia = () => {
            this._lascia = null;
            return lascia();
          };
          /* Nel frattempo i timer possono essere finiti tutti. */
          if (!this._scadenze.size) this._ascolta();
        },
        (errore) => {
          this._abbonandomi = null;
          this.registro.attenzione(`non seguo gli stati del clima: ${errore?.message || errore}`);
        },
      );
  }

  _cambiata(evento) {
    const dati = evento?.data;
    const entita = dati?.entity_id;
    if (!entita || !this._scadenze.has(entita)) return;
    const nuovo = dati.new_state;
    /* `null` e' l'entita' sparita, non spenta: un'integrazione che si ricarica
     * non deve buttare il timer di chi sta dormendo. */
    if (!nuovo || nuovo.state === "unknown" || nuovo.state === "unavailable") return;
    if (nuovo.state === "off") this.annulla(entita);
  }

  /* Spegne per bene: gli allarmi si disarmano (il file li ricorda) e quello
   * che era da scrivere si scrive adesso. */
  chiudi() {
    for (const entita of [...this._allarmi.keys()]) this._disarma(entita);
    if (this._lascia) {
      const lascia = this._lascia;
      this._lascia = null;
      Promise.resolve()
        .then(() => lascia())
        .catch(() => {});
    }
    this.scriviAdesso();
  }
}

File: officina/custom_components/dashboardmodern/spegnimento.py
"""Lo spegnimento programmato del clima: il conto alla rovescia sta in casa.

«Vorrei uno slider per decidere il tempo che un condizionatore debba restare
acceso dal momento che gli do l'on, utile spesso di notte o per accensioni a
spot» (#364).

La parte che conta non e' lo slider, e' **dove vive il conto alla rovescia**.
Un timer nel browser muore chiudendo la pagina, e chi accende il condizionatore
per due ore prima di dormire la pagina la chiude sempre: resterebbe acceso tutta
la notte, cioe' esattamente il contrario di quello che ha chiesto. Il timer
quindi lo tiene Home Assistant.

Tre scelte reggono il file.

* **La scadenza si scrive sul disco.** Un riavvio di Home Assistant nel mezzo
  della notte non deve lasciare acceso un condizionatore: all'avvio le scadenze
  si rileggono e si riarmano. Quelle gia' passate mentre era spento si eseguono
  subito — meglio spegnere in ritardo che non spegnere.
* **Lo spegnimento e' `homeassistant.turn_off`.** Un clima si spegne con
  `climate.turn_off`, ma nella casella dell'unita' puo' esserci anche uno
  `switch` o un `input_boolean` — e chiamare il servizio sbagliato vuol dire
  un timer che scade e non fa niente. `homeassistant.turn_off` sceglie da se'
  il dominio giusto, che e' la stessa cosa che fa il resto della plancia.
* **Il timer si annulla da solo quando l'unita' viene spenta a mano.** Chi
  spegne prima non deve ritrovarsi un timer appeso che, due ore dopo, spegne
  un'unita' che nel frattempo qualcun altro aveva riacceso.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .const import DOMAIN

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant

STORAGE_KEY = f"{DOMAIN}.spegnimenti"
STORAGE_VERSION = 1
DATA_SPEGNIMENTO = "spegnimento"

#: Il piu' corto e il piu' lungo che si possano chiedere, in minuti. Sono gli
#: stessi estremi dello slider in plancia: un valore fuori scala e' un errore
#: di chi chiama, non una richiesta da esaudire.
MIN_MINUTI = 1
MAX_MINUTI = 720

#: Quanti timer possono stare appesi insieme. Uno per unita' climatica, e le
#: unita' sono una decina: cento e' un fondo, non un limite che si tocca.
MAX_TIMER = 100


def _adesso(hass: HomeAssistant) -> float:
    """L'ora di Home Assistant, in millisecondi come la plancia li conta."""
    from homeassistant.util import dt as dt_util

    return dt_util.utcnow().timestamp() * 1000


class SpegnimentoStore:
    """Le scadenze appese, e i loro allarmi."""

    def __init__(self, hass: HomeAssistant) -> None:
        """Aggancia lo store al deposito di Home Assistant."""
        from homeassistant.helpers.storage import Store

        self.hass = hass
        self._store: Store[dict[str, Any]] = Store(hass, STORAGE_VERSION, STORAGE_KEY)
        #: entity_id -> scadenza in millisecondi.
        self._scadenze: dict[str, float] = {}
        #: entity_id -> la funzione che disarma l'allarme.
        self._allarmi: dict[str, Any] = {}
        #: La funzione che smette di seguire gli stati, quando c'e' da seguire.
        self._ascolto: Any = None
        self._caricato = False

    # ── memoria ─────────────────────────────────────────────────────────────

    async def async_load(self) -> None:
        """Rileggi le scadenze e rimetti in piedi gli allarmi."""
        if self._caricato:
            return
        self._caricato = True
        stored = await self._store.async_load()
        if isinstance(stored, dict):
            grezze = stored.get("scadenze")
            if isinstance(grezze, dict):
                for entita, scadenza in grezze.items():
                    if not isinstance(entita, str) or "." not in entita:
                        continue
                    try:
                        self._scadenze[entita] = float(scadenza)
                    except (TypeError, ValueError):
                        continue
        for entita in list(self._scadenze):
            self._arma(entita)
        self._ascolta()

    def _salva(self) -> None:
        """Scrivi con ritardo: una raffica di tocchi non e' una raffica di file."""
        self._store.async_delay_save(lambda: {"scadenze": dict(self._scadenze)}, 5)

    # ── allarmi ─────────────────────────────────────────────────────────────

    def _disarma(self, entita: str) -> None:
        """Togli l'allarme di quell'entita', se ce n'e' uno."""
        annulla = self._allarmi.pop(entita, None)
        if annulla is not None:
            annulla()

    def _arma(self, entita: str) -> None:
        """Metti l'allarme per la scadenza di quell'entita'.

        Una scadenza gia' passata — il riavvio nel mezzo della notte — non si
        butta: si esegue subito. Meglio spegnere in ritardo che non spegnere.
        """
        from homeassistant.helpers.event import async_call_later

        self._disarma(entita)
        scadenza = self._scadenze.get(entita)
        if scadenza is None:
            return
        fra = max(0.0, (scadenza - _adesso(self.hass)) / 1000)

        async def _scatta(_ora: Any) -> None:
            self._allarmi.pop(entita, None)
            await self.async_spegni(entita)

        self._allarmi[entita] = async_call_later(self.hass, fra, _scatta)

    # ── quello che la plancia chiede ────────────────────────────────────────

    def scadenze(self) -> dict[str, float]:
        """Le scadenze appese, cosi' come stanno."""
        return dict(self._scadenze)

    async def async_programma(self, entita: str, minuti: int) -> float | None:
        """Programma lo spegnimento di quell'entita' fra quei minuti.

        Zero minuti vuol dire «togli il timer»: e' come si annulla, e torna
        ``None`` perche' non c'e' nessuna scadenza da mostrare.
        """
        await self.async_load()
        if minuti <= 0:
            return await self.async_annulla(entita)
        minuti = max(MIN_MINUTI, min(int(minuti), MAX_MINUTI))
        if entita not in self._scadenze and len(self._scadenze) >= MAX_TIMER:
            # Il fondo: prima cede la scadenza piu' lontana, che e' quella che
            # ha ancora piu' tempo per essere rimessa.
            piu_lontana = max(self._scadenze, key=self._scadenze.__getitem__)
            self._disarma(piu_lontana)
            self._scadenze.pop(piu_lontana, None)
        self._scadenze[entita] = _adesso(self.hass) + minuti * 60_000
        self._arma(entita)
        self._salva()
        self._ascolta()
        return self._scadenze[entita]

    async def async_annulla(self, entita: str) -> None:
        """Togli il timer di quell'entita', senza toccare l'entita'."""
        await self.async_load()
        self._disarma(entita)
        if self._scadenze.pop(entita, None) is not None:
            self._salva()
            self._ascolta()
        return None

    async def async_spegni(self, entita: str) -> None:
        """La scadenza e' arrivata: spegni, e togli il timer.

        `homeassistant.turn_off` sceglie il dominio giusto da se': nella casella
        dell'unita' puo' esserci un `climate`, ma anche uno `switch` o un
        `input_boolean`, e un servizio sbagliato sarebbe un timer che scade
        senza spegnere niente.
        """
        self._disarma(entita)
        c_era = self._scadenze.pop(entita, None) is not None
        if c_era:
            self._salva()
            self._ascolta()
        stato = self.hass.states.get(entita)
        if stato is None:
            return
        await self.hass.services.async_call(
            "homeassistant", "turn_off", {"entity_id": entita}, blocking=False
        )

    # ── chi spegne a mano si porta via anche il timer ───────────────────────

    def _ascolta(self) -> None:
        """Segui le entita' che hanno un timer, e nessun'altra.

        Senza questo restava appeso un allarme che, due ore dopo, spegneva
        un'unita' che nel frattempo qualcun altro aveva riacceso: il timer
        appartiene a QUELLA accensione, e con lei finisce.
        """
        from homeassistant.helpers.event import async_track_state_change_event

        if self._ascolto is not None:
            self._ascolto()
            self._ascolto = None
        if not self._scadenze:
            return

        async def _cambiata(evento: Any) -> None:
            entita = evento.data.get("entity_id")
            nuovo_stato = evento.data.get("new_state")
            if entita not in self._scadenze:
                return
            # `None` e' l'entita' sparita, non spenta: un'integrazione che si
            # ricarica non deve buttare il timer di chi sta dormendo.
            if nuovo_stato is None or nuovo_stato.state in ("unknown", "unavailable"):
                return
            if nuovo_stato.state == "off":
                await self.async_annulla(entita)

        self._ascolto = async_track_state_change_event(
            self.hass, list(self._scadenze), _cambiata
        )


async def async_get_spegnimento_store(hass: HomeAssistant) -> SpegnimentoStore:
    """Lo store di questa installazione, caricato una volta sola."""
    domain_data: dict[str, Any] = hass.data.setdefault(DOMAIN, {})
    store = domain_data.get(DATA_SPEGNIMENTO)
    if store is None:
        store = SpegnimentoStore(hass)
        domain_data[DATA_SPEGNIMENTO] = store
    await store.async_load()
    return store
